components/codec/src/bits/buffered_read.rs

```diff
@@ -0,0 +1,224 @@
+// Copyright 2023 The CeresDB Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::bits::{Bit, Error};
+
+/// BufferedReader
+///
+/// BufferedReader encapsulates a buffer of bytes which can be read from.
+#[derive(Debug)]
+pub struct BufferedReader<'a> {
+    bytes: &'a [u8], // internal buffer of bytes
+    index: usize,    // index into bytes
+    pos: u32,        // position in the byte we are currently reading
+}
+
+impl<'a> BufferedReader<'a> {
+    /// new creates a new `BufferedReader` from `bytes`
+    pub fn new(bytes: &'a [u8]) -> Self {
+        BufferedReader {
+            bytes,
+            index: 0,
+            pos: 0,
+        }
+    }
+
+    fn get_byte(&mut self) -> Result<u8, Error> {
+        self.bytes.get(self.index).cloned().ok_or(Error::Eof)
+    }
+}
+
+impl<'a> BufferedReader<'a> {
+    pub fn read_bit(&mut self) -> Result<Bit, Error> {
+        if self.pos == 8 {
+            self.index += 1;
+            self.pos = 0;
+        }
+
+        let byte = self.get_byte()?;
+
+        let bit = if byte & 1u8.wrapping_shl(7 - self.pos) == 0 {
```

Comment:
I guess define a global static mask array will provide a better performance:
```
let BIT_MASKS: [u8;8] = [1, 2, 4, 8, 16, ..., 128];
```

With the BIT_MASKS, we can put it here:
```suggestion
        let bit = if byte & BIT_MASK[self.pos] == 0 {
```